File: recsys-pipeline/services/spark-streaming-job/src/main/scala/com/demo/sequence/SequenceCodec.scala
package com.demo.sequence

/** Pure manipulation of packed column strings. Every alignment rule in the store lives
  * here so it can be tested without Spark or Redis. */
object SequenceCodec {

  def pack(values: Seq[String]): String =
    values.map(v => if (v == null) "" else v).mkString(SequenceSchema.RowSeparator)

  /** Split into exactly `n` elements: pad short columns, truncate long ones.
    * The `-1` limit is required — the default drops trailing empty strings. */
  def unpack(packed: String, n: Int): Seq[String] = {
    if (n <= 0) return Seq.empty
    val parts =
      if (packed == null || packed.isEmpty) Array.empty[String]
      else packed.split(SequenceSchema.RowSeparator, -1)
    if (parts.length == n) parts.toSeq
    else if (parts.length > n) parts.take(n).toSeq
    else parts.toSeq ++ Seq.fill(n - parts.length)("")
  }

  /** Concatenate `fresh` after `existing`, keeping the newest `maxRows` rows.
    * Columns absent from either side are treated as all-null so they stay aligned. */
  def merge(
      existing: Map[String, String],
      fresh: Map[String, String],
      maxRows: Int
  ): Map[String, String] = {
    val existingCount = count(existing)
    val freshCount    = count(fresh)
    val total         = existingCount + freshCount
    if (total <= 0) return Map(SequenceSchema.ColCount -> "0")

    val kept = math.min(total, math.max(0, maxRows))
    val drop = total - kept

    val columns = SequenceSchema.Columns.map { column =>
      val combined =
        unpack(existing.getOrElse(column, ""), existingCount) ++
          unpack(fresh.getOrElse(column, ""), freshCount)
      column -> pack(combined.drop(drop))
    }.toMap

    columns + (SequenceSchema.ColCount -> kept.toString)
  }

  private def count(chunk: Map[String, String]): Int =
    chunk.get(SequenceSchema.ColCount)
      .flatMap(v => scala.util.Try(v.trim.toInt).toOption)
      .filter(_ > 0)
      .getOrElse(0)
}

File: recsys-pipeline/services/java-retrieval-service/src/main/java/com/demo/retrieval/service/sequence/SequenceCodec.java
package com.demo.retrieval.service.sequence;

import java.util.ArrayList;
import java.util.List;
import java.util.Map;

/** Read-side unpacking of the packed column strings. Serving never writes sequences. */
public final class SequenceCodec {

    private SequenceCodec() {
    }

    /**
     * Split into exactly {@code n} elements: pad short columns, truncate long ones.
     * The {@code -1} limit is required — the default drops trailing empty strings, which
     * would shift every row whose last value is null.
     */
    public static List<String> unpack(String packed, int n) {
        if (n <= 0) {
            return List.of();
        }
        String[] parts = (packed == null || packed.isEmpty())
            ? new String[0]
            : packed.split(SequenceSchemaConstants.ROW_SEPARATOR, -1);

        List<String> out = new ArrayList<>(n);
        for (int i = 0; i < n; i++) {
            out.add(i < parts.length ? parts[i] : "");
        }
        return List.copyOf(out);
    }

    /**
     * Authoritative row count for a chunk: the {@code n} field, clamped down to the shortest
     * present data column. A torn write must produce fewer rows, never mis-aligned ones.
     */
    public static int rowCount(Map<String, String> fields) {
        int declared = parseCount(fields.get(SequenceSchemaConstants.COL_COUNT));
        if (declared <= 0) {
            return 0;
        }
        int shortest = declared;
        for (String column : SequenceSchemaConstants.COLUMNS) {
            String packed = fields.get(column);
            if (packed == null) {
                continue;   // column absent entirely — nothing to clamp against
            }
            // NOTE: no empty-string special case. A packed "" encodes exactly ONE
            // empty element (a single null value); zero elements is expressed by
            // n == 0, never by the string. Treating "" as length 0 would silently
            // drop every single-event chunk that has a null field — which is the
            // normal shape for a click event with no rating.
            int length = packed.split(SequenceSchemaConstants.ROW_SEPARATOR, -1).length;
            shortest = Math.min(shortest, length);
        }
        return Math.max(0, shortest);
    }

    private static int parseCount(String raw) {
        if (raw == null) {
            return 0;
        }
        try {
            return Integer.parseInt(raw.trim());
        } catch (NumberFormatException e) {
            return 0;
        }
    }
}
